tools/autograd/templates/Functions.cpp

```diff
@@ -726,20 +726,10 @@ Tensor _fused_dropout_backward(Tensor grad, Tensor mask, double p1m) {
   }
 }
 
-Tensor select_first_equal_backward(Tensor grad, const Tensor & input, const Tensor & value) {
-  auto grad_input = at::zeros_like(input);
-
-  // find indices of the first element for which input[idx] == value
-  auto first_value_idx = (input == value).nonzero().select(0, 0);
-
-  if (grad_input.dim() == 0) {
-    grad_input.copy_(grad);
-  }
-  else {
-    grad_input.index_put_(at::chunk(first_value_idx, grad_input.dim()), grad);
-  }
-
-  return grad_input;
+Tensor evenly_dispatch_backward(Tensor grad, const Tensor & input, const Tensor & value) {
+  auto mask = (input == value);
+  auto count = mask.sum(input.scalar_type());
+  return at::zeros_like(input).masked_fill_(mask, grad / count);
```

Comment:
I wonder if it would be faster to do `mask.to(input.scalar_type()) * (grad / count)` here?
Also it might be worth it to special case when `count=1` where we would be able to do something more efficient than `masked_fill_()`.

Reply:
Good point, not even `mask.to` because TensorIterator would take care of it, and on cuda type promotion will be implicit. `grad/count` has to be converted to input.scalar_type for this to work though.

Reply:
Whether to choose `*` vs `masked_fill` depend on CPU vs CUDA:

```python
import torch
for device in ['cpu', 'cuda']:
    t = torch.torch.randint(100, (1024, 1024, 64), device=device, dtype=torch.float)
    s = t.sum()
    m = t.max()
    mask = (t == m)
    go = t.new_tensor(1.)
    torch.cuda.synchronize()
    print(device)
    %timeit mask * (go / s); torch.cuda.synchronize() if device == 'cuda' else None
    %timeit torch.zeros_like(t).masked_fill_(mask, go / s); torch.cuda.synchronize() if device == 'cuda' else None
```
```
cpu
44.5 ms ± 199 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
25.2 ms ± 130 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
cuda
848 µs ± 16.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.08 ms ± 19.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
so there should be some logic
```
if (cuda) {
  *
} else {
  masked_fill
}
```
and I don't think it should have a separate case for `s == 1`, because the only difference is `grad / count` vs `grad`, and `grad` is a scalar tensor. `scalar_tensor.item()` is not faster than `scalar_tensor / another_scalar_tensor`.